packages/react-dom/src/fire/ReactFireEventTraversal.js

```diff
@@ -0,0 +1,234 @@
+  // Html Nodes can be nested fe: span inside button in that scenario
+  // browser does not handle disabled attribute on parent,
+  // because the event listener is on document.body
+  // Don't process clicks on disabled elements
+  if (proxyContext.isClickEvent && domNode.disabled) {
+    return;
+  }
+  const domNodeEventsMap = getDomNodeEventsMap(domNode);
+  const {event, eventName} = proxyContext;
+  triggerEventHandler(
+    domNode,
+    domNodeEventsMap,
+    capturePhase ? `${eventName}-capture` : eventName,
+    proxyContext,
+    event,
+  );
+  if (event.cancelBubble) {
+    return true;
+  }
+}
+
+export function getEventTargetAncestorFibers(targetFiber) {
+  const ancestors = [];
+  // Loop through the hierarchy, in case there's any nested components.
+  // It's important that we build the array of ancestors before calling any
+  // event handlers, because event handlers can modify the DOM, leading to
+  // inconsistencies with ReactMount's node cache. See #1105.
+  let ancestor = targetFiber;
+  do {
+    if (!ancestor) {
+      break;
+    }
+    const root = findRootContainerNode(ancestor);
+    if (!root) {
+      break;
+    }
+    ancestors.push(ancestor);
+    ancestor = getClosestFiberFromDOMNode(root);
+  } while (ancestor);
+  return ancestors;
+}
+
+export function traverseTwoPhase(proxyContext) {
+  let {fiber} = proxyContext;
+  let i;
+  let stopped;
+  const path = [];
+  while (fiber != null) {
+    path.push(fiber);
+    fiber = getParent(fiber);
+  }
+
+  for (i = path.length; i-- > 0; ) {
+    stopped = dispatchEventHandler(path[i], true, proxyContext);
+    if (stopped) {
+      return;
+    }
+  }
+  for (i = 0; i < path.length; i++) {
+    stopped = dispatchEventHandler(path[i], false, proxyContext);
+    if (stopped) {
+      return;
+    }
+  }
+}
+
+/**
+ * Return the lowest common ancestor of A and B, or null if they are in
+ * different trees.
+ */
+export function getLowestCommonAncestor(instA, instB) {
+  let depthA = 0;
+  for (let tempA = instA; tempA; tempA = getParent(tempA)) {
+    depthA++;
+  }
+  let depthB = 0;
+  for (let tempB = instB; tempB; tempB = getParent(tempB)) {
+    depthB++;
+  }
+
+  // If A is deeper, crawl up.
+  while (depthA - depthB > 0) {
+    instA = getParent(instA);
+    depthA--;
+  }
+
+  // If B is deeper, crawl up.
+  while (depthB - depthA > 0) {
+    instB = getParent(instB);
+    depthB--;
+  }
+
+  // Walk in lockstep until we find a match.
+  let depth = depthA;
+  while (depth--) {
+    if (instA === instB || instA === instB.alternate) {
+      return instA;
+    }
+    instA = getParent(instA);
+    instB = getParent(instB);
+  }
+  return null;
+}
+
+/**
+ * Traverses the ID hierarchy and invokes the supplied `cb` on any IDs that
+ * should would receive a `mouseEnter` or `mouseLeave` event.
+ *
+ * Does not invoke the callback on the nearest common ancestor because nothing
+ * "entered" or "left" that element.
+ */
+export function traverseEnterLeave(
```

Comment:
would this be a good moment/save implementation complexity to change up how this plugin is implemented, could avoid the special traversal

Reply:
Do you have any ideas how we can change this plugin? For now the idea was to leave the current logic mostly alone and try and refactor with the new codebase operational. It would be good if you have some ideas how to avoid doing all this extra work.